### DISCUSSION TITLE
Rounded up media channel spend values in optimize.optimized_data output

### DISCUSSION THREAD

**Discussion by RagarshV** [google/meridian#378] [Q&A] [answered]
The optimized_data function returns rounded optimized channel spend numbers. Ideally, we would want the budget allocator to optimize towards the provided budget to the exact dollar, the rounding up inflates the total optimized budget. 

Is there a way get optimized spend output without rounding up optimized channel spend values? 

To give an example - 

Considering historical spend is $1,25,263

optimization = optimizer.BudgetOptimizer(meridian).optimize(
          selected_times =selected_times,
          budget = 125263,
          pct_of_spend = pct_of_spend,
          spend_constraint_lower = lower_constraints,
          spend_constraint_upper = upper_constraints
        )

the total spend column in optimization.optimized_data() is rounded up at a channel to $1,27,000. ~$2k more than provided budget

**Comment by viktoriias** [comment]
What is your `pct_of_spend`?

**Reply by viktoriias** [reply]
You can change rounding decimal using `gtol`.

**Reply by RagarshV** [reply]
Hi @viktoriias thank you for this, changing `gtol` does get the spend values closer to historical spend. The `pct_of_spend` is defaulted to historical spend. 

I also notice the total incremental_impact in the optimization.optimized_data() output is 6 times the historical KPI. Any insight  into why this could be happening? 

Example of what I am seeing - 
While historical kpi for the optimization period is 58,000, the total `incremental_impact` shown in `optimizer.BudgetOptimizer(meridian).optimize()` for the same period is 369,380 which is 6.3 times the historical, which is too good to be true.

**Reply by alexwakim** [reply]
I would first make sure that the model has converged. I would then assess model fit and interpret the model visualization. Assuming you’ve done this, it’s still possible that the `BudgetOptimzer` produces extreme results, particularly under certain settings.

Consider the `spend_constraint_lower` and `spend_contraint_upper` arguments in `BudgetOptimizer.optimize`, which constrain the budget shift of each channel (default is plus-or-minus 30%). The larger the budget shift, the larger the difference between historical KPI and expected KPI under optimal budget. Also consider the `use_optimal_frequency` argument in `BudgetOptimizer.optimize`, if you are modeling a channel with reach and frequency (R&F) data. This argument defaults to `True`, which assumes each channel with R&F data is executed at the optimal frequency. If optimal frequency is very different from historical frequency, then this can lead to big differences between historical KPI and expected KPI under optimal budget.

**Reply by RagarshV** [reply]
Thank you @alexwakim, a lot of my settings are within the ranges you mentioned - the model is a National model, it has converged and the model fit r_squared is 85%. The only difference is that `spend_constraint_lower` and `spend_contraint_upper` range between +-10% to +-50% and I am not using R&F data to train the model. 

Is there anything else I am not considering? Is it important to include R&F data to minimize the difference?

**Reply by alexwakim** [reply]
Do the ROI estimates and the response curves look reasonable to you?

**Reply by RagarshV** [reply]
the response curves do not look reasonable, the `incremental_impact` scale at a media_channel level is higher than historical.

**Reply by alexwakim** [reply]
If you set the budget constraints to 30% (default), do you still see big differences between historical KPI and expected KPI under optimal budget?

**Reply by RagarshV** [reply]
Yeah I am still seeing a big difference with the 30% default. While the total for `ci_lo` is around the historical KPI range, the mean is really high although I am using the default 0.9 `confidence_interval`

**Reply by alexwakim** [reply]
> the response curves do not look reasonable

The response curve shows each channel’s incremental outcome at increased levels of spend. Thus it is important for the budget allocator. The first step here is to make sure you have reasonable response curves. Some considerations,

- It can be difficult to correctly infer the response curve for ranges of media execution beyond what’s observed in the data. If a channel has low variation in the `media` argument of `InputData`, it may be the case that the budget allocator is going beyond what’s observed in the data. Prevent this by decreasing the budget constraints for any such channels to a value that ensures budget allocation doesn't go beyond what's observed in the data.
- Consider a custom [`ec_m` prior](https://developers.google.com/meridian/docs/advanced-modeling/default-prior-distributions#ec_m_and_ec_om) that encourages more saturated response curves. A prior with more mass at smaller values will encourage this.
- Consider setting the [paid_media_prior_type](https://github.com/google/meridian/blob/7b5f0ad12c55db4de4c2be90e41c1a7de957e6ce/meridian/model/spec.py#L75) argument to `'mroi'`. A more informative prior on mROI will shrink posterior optimal budget allocation towards historical.

**Reply by RagarshV** [reply]
Thank you for the considerations @alexwakim. Tested the approaches you suggested and it doesn't seem to make a big difference in truncating the response curves. 

- To your first point, the `InputData` I am using goes back to around two and a half years with sufficient variance across media channels. 
- The custom [ec_m prior](https://developers.google.com/meridian/docs/advanced-modeling/default-prior-distributions#ec_m_and_ec_om) did help truncate the scale of the response curves by a little bit, but the volume is still higher than historical.
- I have been modeling towards kpi so setting [paid_media_prior_type](https://github.com/google/meridian/blob/7b5f0ad12c55db4de4c2be90e41c1a7de957e6ce/meridian/model/spec.py#L75) argument to 'mroi'. might not make sense.

**Reply by yingxzgo** [reply, marked as answer]
Thanks for the additional info, Here are our suggested steps to solve the issue:

1. To begin, please check if the model has a negative baseline KPI ([Code](https://developers.google.com/meridian/docs/user-guide/plot-media-visualizations#waterfall-chart)). A negative baseline indicates that the model's estimated incremental KPI due to paid media surpasses the historical KPI. Further explanation and solutions regarding this issue can be found here ([link](https://developers.google.com/meridian/docs/advanced-modeling/model-debugging#baseline-low-or-negative)). If the baseline is positive, proceed to Step 2 & 3.

2. Consider identifying which channels are receiving budget increases in the optimization. These channels must also have high marginal ROI estimates (much higher than the historical Total Media ROI estimates).

- Note that Meridian has changed the name `incremental_impact` to `incremental_outcome`.
- In your case, the outcome is KPI not revenue, so Meridian defines ROI and marginal ROI in terms of incremental KPI instead of revenue.

.3. Once you have identified the channels with largest marginal ROI, here are some suggestions:

- Modify the prior distribution on these channels. Shrinking the ROI or EC parameters toward zero for these specific channels will likely reduce the marginal ROI estimates for these channels.
- Apply stronger optimization constraints on these channels to prevent them from receiving too much budget allocation under optimization.
